flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/AbstractUserClassPathJobGraphRetriever.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.entrypoint.component;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.FileVisitOption;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ *  Abstract class for the JobGraphRetriever, which wants to get classpath user's code depends on.
+ */
+public abstract class AbstractUserClassPathJobGraphRetriever implements JobGraphRetriever {
+
+	protected static final Logger LOG = LoggerFactory.getLogger(AbstractUserClassPathJobGraphRetriever.class);
+
+	/** The directory contains all the jars, which user code depends on. */
+	@Nullable
+	private final String jobDir;
+
+	private List<URL> userClassPaths;
+
+	public AbstractUserClassPathJobGraphRetriever(String jobDir) {
+		this.jobDir = jobDir;
+	}
+
+	protected List<URL> getUserClassPaths() throws IOException {
+		if (userClassPaths == null) {
+			userClassPaths = getRelativeJarsURLFromDir(jobDir);
+		}
+		return userClassPaths;
+	}
+
+	/**
+	 * Scan all the jar files in the {@code dir} and return all these jar files' relative URLs to "user.dir".
+	 * @param dir the dir needed to scan the jar files
+	 * @return the jar files' relative URLs
+	 * @throws IOException
+	 */
+	private List<URL> getRelativeJarsURLFromDir(String dir) throws IOException {
```

Comment:
This method is static and, hence, does not really need to be part of this class.

Reply:
Is this method supposed to find jars on a local directory or an arbitrary path? I would suggest to change the type of the parameter accordingly (`File` vs. `Path`).

Reply:
```suggestion
	private List<URL> listJarsInDirectory(String dir) throws IOException {
```

Reply:
> This method is static and, hence, does not really need to be part of this class.

I think I could move this method to a Util class  for example `JobGraphRetrieverUtil`

> Is this method supposed to find jars on a local directory or an arbitrary path? I would suggest to change the type of the parameter accordingly (`File` vs. `Path`).

Currently, this implementation implies that scan the jar in  the local directory.(`new File(job).toPath()` and `Paths.get(job)` both use the same `FileSystems.getDefault()`)

However, I think this method could make no assumption about the `FileSystem`.  Let the caller decide what FileSystem to use. For example `StandaloneJobClusterEntryPoint` decides to use the 'local' FileSystem.

So I prefer to use the `Path`

Reply:
I'm fine with `Path`.

What about moving creating a `FileUtils#listFilesInPath(Path directory, Predicate fileFilter)` for the file listing?

Reply:
After the offline discussion with Till. This method will be breaking down to three functions
`List<File> listFilesInPath(File directory, Predicate fileFilter)`
`List<File> relativizeToWorkingDir(List<File>)`
`List<URL> toRelativeURL(List<File>)`

Reply:
ok I will fix it